### ISSUE TITLE
Add rez-build '--local-only' arg

### ISSUE THREAD

**Issue by nerdvegas** [nerdvegas/rez#652]
Add an option to only see local packages, if a given package family has any local packages available.

Currently if you have local 'foo-1.0.0' available, but released 'foo-1.0.1' is also visible, then foo-1.0.1 is still taken as priority, even if you're currently developing foo and didn't expect there to be a higher-versioned released package.

The option should take nargs, in case you only want to enable it for specific packages.

**Comment by bfloch** [comment]
Nice suggestion!
Isn't the wording of **--local-only** miss-leading?
Shouldn't it be more something like: **--prefer-local**

**Comment by JeanChristopheMorinRodeoFX** [comment]
I also like the idea, but I'd prefer the naming suggested by @bfloch .

**Comment by nerdvegas** [comment]
The naming is tricky. Eg, prefer-local suggests that local versions would
be tried first, and non-local second, but that doesn't describe the
functionality (doing that may be pretty complicated - I can't say for sure
without looking into this more closely). With this feature, _only_ local
versions of a package would be used, if any local versions exist. Consider
what the help will look like:

```
--local-only [PKG, ..., PKG]: Use only the local versions of the given
packages, if local packages are present. If PKG is not specified, apply to
all packages.
```

Typically someone would use this because they want to be confident that
their local package is being used, for testing purposes, so with that in
mind I would think that the local-only behaviour would be more desired than
the local-preferred behaviour as described above.

Thoughts?
A

**Comment by bfloch** [comment]
Fair. Especially since it can be used without arguments.

Rethinking it, you are right. Failing is probably the right thing to do if it can't resolve it to the locals, else we will just have default behaviour.

Is there a story where `--no-local` might want to also take arguments? As in: I am working on a bunch of packages locally and want to test against one of them being resolved from the release repo?

**Comment by nerdvegas** [comment]
Yes, I think it would be fair to update --no-local to be able to specify
packages - that would feel more consistent with --local-only added (one is
actually the opposite of the other). That should probably be done in the
same PR that adds --local-only imo.

**Comment by mottosso** [comment]
> update --no-local to be able to specify packages

Throwing some suggestions into the mix.

```bash
$ rez build --exclude-path /home/marcus/packages # i.e. --no-local
$ rez build --include-path /packages             # i.e. --local-only
```

To work akin to..

```bash
$ rez env --exclude
$ rez env --include
```

In that they either explicitly include *only* a set of paths with `--include`, effectively meaning `--local-only` with that path is `$REZ_LOCAL_PACKAGES_PATH`, or for granularity..

```bash
$ rez env --exclude-path $REZ_LOCAL_PACKAGES_PATH/int
```

That way compatibility can be preserved, consistency maintained amongst arguments and an argument to the flag be expected (as opposed to `--no-local` and `--local-only` which is a little unclear about whether or not it takes an argument, grammatically speaking)

**Comment by JeanChristopheMorinPerso** [comment]
Hey @nerdvegas I've thought more on this and your proposal make sense with your extra information you gave after my response.

I think we could also add the same flag inside `rez-test`. If you `rez-build --local-only`, you might also be interested to run the package tests with the exact same dependencies artifacts that your package was built with.

What do you think?

**Comment by nerdvegas** [comment]
Agreed, particularly as rez-test also supports --no-local already.

Related - no-local should also be updated to take an optional list of
packages, so it matches local-only. Should be straightforward for this to
be backwards compatible, just have to support no-local being either a bool,
or list of str (and fortunately the cli form is backwards compatible also).

Cheers
A
